src/IO/PeekableReadBuffer.cpp

```diff
@@ -20,6 +20,19 @@ PeekableReadBuffer::PeekableReadBuffer(ReadBuffer & sub_buf_, size_t start_size_
     checkStateCorrect();
 }
 
+void PeekableReadBuffer::reset()
+{
+    peeked_size = 0;
+    checkpoint = nullptr;
+    checkpoint_in_own_memory = false;
+
+    if (!currentlyReadFromOwnMemory())
+        sub_buf.position() = pos;
```

Comment:
Is it possible that `pos` became an invalid pointer because internal memory of `sub_buf` was reallocated for some reason after direct manipulations with `sub_buf`?

Reply:
All direct manipulations on subbuf possible only after after reset() call before next(). May be i should add comment dicumneting that requirement.

Reply:
Added a comment in PeekableReadBuffer.h